File: google_codejam(2020)/c++/problem_2.cpp
#include<bits/stdc++.h>
#include<string>
using namespace std;
int main(){
    int c;
    cin>>c;
    for(int i=1;i<=c;i++){
        string s,res;
        int prev,pres;
        cin>>s;
        for(int i=0;i<s.length();i++){
            if(res.empty()){
                pres = s[i]-'0';
                for(int j=0;j<pres;j++){
                    res += '(';
                }
                res += s[i];
                prev = pres;
            }else{
                pres = s[i]-'0';
                if(pres==prev){
                    res += s[i];
                }else{
                    if(pres>prev){
                        for(int i=0;i<pres-prev;i++){
                            res += '(';
                        }
                        res += s[i];
                        prev = pres;
                    }else{
                        for(int i=0;i<prev-pres;i++){
                            res += ')';
                        }
                        res += s[i];
                        prev = pres;
                    }
                }
            }
        }
        for(int i=0;i<prev;i++){
            res += ")";
        }
        cout<<"Case #"<<i<<": "<<res<<"\n";
    }
    return 0;
}

File: google_codejam(2020)/python/problem_2.py
case = int(input())
for ca in range(1,case+1):
    s = list(map(int,input()))
    re = []
    prev = 0
    count = 0
    for i in s:
        #if len(res) == 0:
        if i == prev:
            re.append(i)
        elif i > prev:
            re.append("("*(i-prev))
            re.append(i)
            prev = i 
        else:
            re.append(")"*(prev-i))
            re.append(i)
            prev = i    
        count += 1            
        if count == len(s) and prev != 0:
            re.append(")"*prev)   
    res = ""
    for i in re:
        res += str(i)        

    print("Case #{}: {}".format(ca, res))



